Include/internal/pycore_frame.h

```diff
@@ -118,6 +119,7 @@ _PyFrame_InitializeSpecials(
     frame->f_state = FRAME_CREATED;
     frame->is_entry = false;
     frame->is_generator = false;
+    frame->self = NULL;
```

Comment:
Note to reviewers: tied to frame state instead of some cache/call_shape so that subsequent nested calls don't destroy `self` (and we can identify which frame the `self` belongs to). Consider the following code:
```python
class Tokenizer:
    def __init__(self):
        self.__next() # Kaboom!
    def __next(self):
        pass

for _ in range(10):
 print(Tokenizer())
```